src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRefreshListener.cs

```diff
@@ -45,9 +47,27 @@ internal class SemanticTokensRefreshListener
 
         private void OnLspWorkspaceChanged(object? sender, WorkspaceChangeEventArgs e) => _workQueue.AddWork();
 
-        // TO-DO: Replace hardcoded string with const once LSP side is merged.
-        public ValueTask SendSemanticTokensNotificationAsync(CancellationToken cancellationToken)
-            => _languageServerNotificationManager.SendNotificationAsync(Methods.WorkspaceSemanticTokensRefreshName, cancellationToken);
+        public async ValueTask SendSemanticTokensNotificationAsync(CancellationToken cancellationToken)
+        {
+            // Currently, we send a refresh notification to the client whenever there are any workspace changes.
+            // However, we also need to send notifications whenever the compilations finish processing since
+            // compilations impact semantic tokens. We put this logic here rather than in the semantic tokens
+            // range handler so that the client still gets initial colorization on file open.
+            var trackedDocuments = _lspWorkspaceManager.GetTrackedLspText();
+            var projects = trackedDocuments.Select(
+                d => _lspWorkspaceManager.GetLspDocument(new TextDocumentIdentifier { Uri = d.Key })?.Project).Distinct();
+
+            foreach (var project in projects)
+            {
+                if (project is null)
+                    continue;
+
+                cancellationToken.ThrowIfCancellationRequested();
+                await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                await _languageServerNotificationManager.SendNotificationAsync(
```

Comment:
Razor debounces refresh notifications so sending more notifications shouldn't be an issue. @gundermanc just to confirm, does the LSP client do the same?

Reply:
>  @gundermanc just to confirm, does the LSP client do the same?

I don't think we do that currently.

Reply:
Would debouncing on the client be an option for M2? Maybe batching all the requests within a 1-2 second window?

Reply:
I think we should de-bounce on our side regardless of what the client does, either with feeding this as an event from range (as in above comments) or via queuing this compilation available event into the work queue

Reply:
> Would debouncing on the client be an option for M2? Maybe batching all the requests within a 1-2 second window?

I would vote we not take that to the Client for M2 to avoid churn. It's an optimization rather than a "bug fix" (unless I'm missing something) and I would project that while the impact would be measurable it probably wouldn't cause high-level Watsons or anything (famous last words).